contribute-to-tidb/review-a-pr-stage.md

```diff
@@ -0,0 +1,233 @@
+
+### Principles of the code review
+
+* Technical facts and data overrules opinions and personal preferences
+* Software design is about trade-off and there is no silver bullet
+
+Everyone comes from different technical backgrounds with different knowledge. They have their own personal preferences. It is important that the code review is not based on biased opinions.
+
+A common mistake is to focus too much on style. Style is important, that is why there are style guides. Just let the bot do most of the trivial work. The code reviewer is a human, who is supposed to focus on things beyond a bot can do. They must make sure that the logic is not broken by the changes, and the maintenance cost of the code base is not decreased.
+
+Sometimes, making choices can be tricky.
+
+Said that there is a special optimization that can improve the overall performance by 30%. However, it introduces a totally different code path and every feature later on has to consider this code path specially, which means a huge maintenance burden. Shall we accept or reject it?
+
+Said that there is a critical bug and a PR is read to fix it, but the change is risky to introduce other kind of bugs. Shall we accept or reject it?
+
+The answer is always “it depends”. Software design is more like a kind of art than technology. It is about aesthetics and your taste of the code. Trade-offs are always there and there is no perfect solution sometimes. Still, there is a general standard (from [@shenli](https://github.com/shenli), maybe we can call it "better than before?"):
+
+* It is making something better, although not a perfect solution.
+* It can mitigate the pain of the user.
+* There is no bug introduced by the change itself.
+
+Then the code can be accepted.
+
+Of course, code review should have some minimal requirements.
+
+For a bug fix PR, if it comes with no test case, the reviewer should ask the author to add tests.
+
+For a performance improvement PR, if no benchmark result is provided, the reviewer should ask the author to attach a benchmark result.
+
+## What to look for in a code review
+
+### Design
+
+The most important thing to cover in a review is the overall design.
+
+Read the description of the PR, and if it's not easy to understand, ask the author to improve it. You may consider questions like:
+
+* Does the change integrate well with the rest of our system?
+* Is it a good time now to add this functionality?
+
+[TiDB Internals forum](https://internals.tidb.io/) is a place for discussion. For pull requests that are created without prior consensus, try to seek consensus there.
+
+### Functionality
+
+* Does this PR do what the developer intend to do?
+* Is what the developer intend to do good for the users with this implementation?
+
+Basically, PR authors are expected to test their PRs well-enough before code reviews. However, as a reviewer you should still think about edge cases, especially for problems about concurrency.
```

Comment:
... to sufficiently test their PRs before code reviews. However, as a reviewer, you should still think about edge cases -- especially for problems about concurrency.

Reply:
```suggestion
Basically, PR authors are supposed to sufficiently test their PRs before code reviews. However, as a reviewer, you should still think about edge cases, especially for problems about concurrency.
```